### PR TITLE
[`MptConfig`] support from pretrained args

### LINKED ISSUE #25114: MptForCausalLM.from_pretrained gives error 'dict' object has no attribute 'softmax_scale'
### System Info

Creating model MPT:
```
    model = MptForCausalLM.from_pretrained(
        model_args.model_name_or_path,
        cache_dir=training_args.cache_dir,
        torch_dtype=torch.bfloat16,
        use_cache=False,
        init_device=f"cuda:{local_rank}",
        attn_config=dict(attn_impl="flash", softmax_scale=None),  # triton, flash
    )
```
Gives the following error:
```
  File "/home/anton/personal/stanford_alpaca-replit/env/lib/python3.10/site-packages/transformers/models/mpt/modeling_mpt.py", line 258, in __init__
    self.attn = MptAttention(config)
  File "/home/anton/personal/stanford_alpaca-replit/env/lib/python3.10/site-packages/transformers/models/mpt/modeling_mpt.py", line 137, in __init__
    self.softmax_scale = config.attn_config.softmax_scale
AttributeError: 'dict' object has no attribute 'softmax_scale'
```

### Reproduction

```
    model = MptForCausalLM.from_pretrained(
        model_args.model_name_or_path,
        cache_dir=training_args.cache_dir,
        torch_dtype=torch.bfloat16,
        use_cache=False,
        init_device=f"cuda:{local_rank}",
        attn_config=dict(attn_impl="flash", softmax_scale=None),  # triton, flash
    )
```

### Expected behavior

Should be using `MptAttentionConfig` instead is using a dict object

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Adds a setter for `attn_config` to allow passing dict when initializing for backward compatibility 
Fixes https://github.com/huggingface/transformers/issues/25114

### CODE REVIEW

File: src/transformers/models/mpt/configuration_mpt.py [hunk 1]
```diff
@@ -101,6 +101,24 @@ def __init__(
                 f"`attn_type` has to be either `multihead_attention` or `multiquery_attention`. Received: {attn_type}"
             )
 
+    @classmethod
+    def from_pretrained(cls, pretrained_model_name_or_path, **kwargs) -> "PretrainedConfig":
+        cls._set_token_in_kwargs(kwargs)
+
+        config_dict, kwargs = cls.get_config_dict(pretrained_model_name_or_path, **kwargs)
+
+        # get the text config dict if we are loading from CLIPConfig
```
Comment (younesbelkada) [inline src/transformers/models/mpt/configuration_mpt.py:110]:
```suggestion
        # get the text config dict if we are loading from MptConfig
```

**Review by younesbelkada — APPROVED** [review]
Looks great as far as I can tell, thanks for taking this!

File: src/transformers/models/mpt/configuration_mpt.py (hunk 1 above)
Comment (younesbelkada) [inline src/transformers/models/mpt/configuration_mpt.py:110]:
```suggestion
```
actually maybe let's completely remove it

File: src/transformers/models/mpt/configuration_mpt.py [hunk 2]
```diff
@@ -101,6 +101,24 @@ def __init__(
                 f"`attn_type` has to be either `multihead_attention` or `multiquery_attention`. Received: {attn_type}"
             )
 
+    @classmethod
+    def from_pretrained(cls, pretrained_model_name_or_path, **kwargs) -> "PretrainedConfig":
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/mpt/configuration_mpt.py:105]: This is not 100% required, but if we want to init an AttentionConfig using `AttentionConfig.from_pretrained(path-to-mpt)` we need to correctly extract the attn_config. We support this for other composition config like CLIP, might be useful, but also ok to remove

[the remark this answers is not in the document]
Comment (sgugger) [reply src/transformers/models/mpt/configuration_mpt.py:105]: Makes sense.

**Review by sgugger — APPROVED** [review]

**Pull request merged**